File: macOS/Packages/ClaudeUsageCore/Sources/ClaudeUsageCore/SnapshotComputer.swift
import Foundation

/// Turns a flat list of usage events into the displayable snapshot types. Shared by
/// every platform (macOS app, iOS app, and eventually Windows/Android) so "today's
/// total" or "current block" always means exactly the same computation everywhere —
/// this is what makes multi-device sync actually show the same numbers on every
/// device, not just similar-looking ones.
public enum SnapshotComputer {
    public static func computeSnapshot(
        from events: [UsageEvent],
        manualBlockTokenTarget: Double,
        appearance: GaugeAppearance
    ) -> UsageSnapshot {
        let now = Date()
        let calendar = Calendar.current
        let startOfToday = calendar.startOfDay(for: now)
        let sevenDaysAgo = calendar.date(byAdding: .day, value: -7, to: now) ?? now

        let todayEvents = events.filter { $0.timestamp >= startOfToday }
        let last7DaysEvents = events.filter { $0.timestamp >= sevenDaysAgo }
        let last7DaysTotalTokens = last7DaysEvents.reduce(0) { $0 + $1.totalTokens }
        let todayTotalTokens = todayEvents.reduce(0) { $0 + $1.totalTokens }

        var breakdownByModel: [String: ModelBreakdown] = [:]
        for event in todayEvents {
            var entry = breakdownByModel[event.model] ?? ModelBreakdown(
                model: event.model, inputTokens: 0, outputTokens: 0,
                cacheCreationTokens: 0, cacheReadTokens: 0, estimatedCostUSD: nil
            )
            entry.inputTokens += event.inputTokens
            entry.outputTokens += event.outputTokens
            entry.cacheCreationTokens += event.cacheCreationTokens
            entry.cacheReadTokens += event.cacheReadTokens
            breakdownByModel[event.model] = entry
        }

        var todayCost: Double = 0
        // Some transcript lines (e.g. internal "<synthetic>" turns) carry a usage block
        // with every field at 0 — harmless to parse, but not worth showing as a line item.
        let breakdown: [ModelBreakdown] = breakdownByModel.values
            .filter { $0.inputTokens + $0.outputTokens + $0.cacheCreationTokens + $0.cacheReadTokens > 0 }
            .map { entry in
                var entry = entry
                let cost = PricingTable.estimatedCostUSD(
                    model: entry.model,
                    inputTokens: entry.inputTokens,
                    outputTokens: entry.outputTokens,
                    cacheCreationTokens: entry.cacheCreationTokens,
                    cacheReadTokens: entry.cacheReadTokens
                )
                entry.estimatedCostUSD = cost
                todayCost += cost ?? 0
                return entry
            }.sorted { $0.model < $1.model }

        let hourlyTokensToday = hourlyBuckets(from: todayEvents.map { ($0.timestamp, $0.totalTokens) }, calendar: calendar)
        let dailyTokensLast7Days = dailyBuckets(from: last7DaysEvents.map { ($0.timestamp, $0.totalTokens) }, calendar: calendar)
        let todayByProject = projectBuckets(from: todayEvents.map { ($0.projectPath, $0.totalTokens) })

        let blocks = SessionBlockCalculator.computeBlocks(from: events)
        let activeBlock = blocks.last.flatMap { now < $0.end ? $0 : nil }
        let pastBlocks = activeBlock == nil ? blocks : blocks.dropLast()
        let historicalMaxBlockTokens = pastBlocks.map(\.totalTokens).max()

        let blockSummary = activeBlock.map {
            SessionBlockSummary(start: $0.start, end: $0.end, totalTokens: $0.totalTokens)
        }

        return UsageSnapshot(
            generatedAt: now,
            todayTotalTokens: todayTotalTokens,
            todayEstimatedCostUSD: todayCost,
            todayModelBreakdown: breakdown,
            todayByProject: todayByProject,
            currentBlock: blockSummary,
            last7DaysTotalTokens: last7DaysTotalTokens,
            hourlyTokensToday: hourlyTokensToday,
            dailyTokensLast7Days: dailyTokensLast7Days,
            historicalMaxBlockTokens: historicalMaxBlockTokens,
            historicalBlockCount: pastBlocks.count,
            manualBlockTokenTarget: manualBlockTokenTarget,
            appearance: appearance
        )
    }

    public static func computeCodexSnapshot(
        from events: [CodexUsageEvent],
        primaryWindow: CodexRateLimitWindow?,
        secondaryWindow: CodexRateLimitWindow?,
        colorHex: String
    ) -> CodexSnapshot {
        let now = Date()
        let calendar = Calendar.current
        let startOfToday = calendar.startOfDay(for: now)
        let sevenDaysAgo = calendar.date(byAdding: .day, value: -7, to: now) ?? now

        let todayEvents = events.filter { $0.timestamp >= startOfToday }
        let last7DaysEvents = events.filter { $0.timestamp >= sevenDaysAgo }
        let last7DaysTotalTokens = last7DaysEvents.reduce(0) { $0 + $1.totalTokens }
        let todayTotalTokens = todayEvents.reduce(0) { $0 + $1.totalTokens }

        var breakdownByModel: [String: ModelBreakdown] = [:]
        for event in todayEvents {
            var entry = breakdownByModel[event.model] ?? ModelBreakdown(
                model: event.model, inputTokens: 0, outputTokens: 0,
                cacheCreationTokens: 0, cacheReadTokens: 0, estimatedCostUSD: nil
            )
            entry.inputTokens += event.inputTokens
            entry.outputTokens += event.outputTokens
            entry.cacheCreationTokens += event.cachedInputTokens
            breakdownByModel[event.model] = entry
        }
        // Codex is a flat subscription (Plus/Pro/etc.), not pay-per-token — no cost column.
        let breakdown = breakdownByModel.values
            .filter { $0.inputTokens + $0.outputTokens > 0 }
            .sorted { $0.model < $1.model }

        let hourlyTokensToday = hourlyBuckets(from: todayEvents.map { ($0.timestamp, $0.totalTokens) }, calendar: calendar)
        let dailyTokensLast7Days = dailyBuckets(from: last7DaysEvents.map { ($0.timestamp, $0.totalTokens) }, calendar: calendar)
        let todayByProject = projectBuckets(from: todayEvents.map { ($0.projectPath, $0.totalTokens) })

        return CodexSnapshot(
            generatedAt: now,
            todayTotalTokens: todayTotalTokens,
            todayModelBreakdown: breakdown,
            todayByProject: todayByProject,
            hourlyTokensToday: hourlyTokensToday,
            dailyTokensLast7Days: dailyTokensLast7Days,
            last7DaysTotalTokens: last7DaysTotalTokens,
            primaryWindow: primaryWindow,
            secondaryWindow: secondaryWindow,
            colorHex: colorHex
        )
    }

    private static func hourlyBuckets(from events: [(timestamp: Date, tokens: Int)], calendar: Calendar) -> [HourlyUsagePoint] {
        var totalsByHour: [Date: Int] = [:]
        for event in events {
            let hourStart = calendar.date(
                from: calendar.dateComponents([.year, .month, .day, .hour], from: event.timestamp)
            ) ?? event.timestamp
            totalsByHour[hourStart, default: 0] += event.tokens
        }
        return totalsByHour
            .map { HourlyUsagePoint(hourStart: $0.key, tokens: $0.value) }
            .sorted { $0.hourStart < $1.hourStart }
    }

    private static func dailyBuckets(from events: [(timestamp: Date, tokens: Int)], calendar: Calendar) -> [DailyUsagePoint] {
        var totalsByDay: [Date: Int] = [:]
        for event in events {
            let day = calendar.startOfDay(for: event.timestamp)
            totalsByDay[day, default: 0] += event.tokens
        }
        return totalsByDay
            .map { DailyUsagePoint(day: $0.key, tokens: $0.value) }
            .sorted { $0.day < $1.day }
    }

    private static func projectBuckets(from events: [(projectPath: String, tokens: Int)]) -> [ProjectUsageEntry] {
        var totalsByProject: [String: Int] = [:]
        for event in events {
            totalsByProject[event.projectPath, default: 0] += event.tokens
        }
        return totalsByProject
            .map { ProjectUsageEntry(projectPath: $0.key, tokens: $0.value) }
            .sorted { $0.tokens > $1.tokens }
    }
}

File: TokenMihariban/Logic/SnapshotComputer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using TokenMihariban.Models;

namespace TokenMihariban.Logic;

/// <summary>
/// Turns a flat list of usage events into the displayable snapshot types — ported from
/// the Mac/iOS/Android `SnapshotComputer` so "today's total"/"current block" mean
/// exactly the same thing on every platform. Day/hour bucketing uses the device's local
/// timezone (`Calendar.current` on the Mac/iOS side), matching that behavior here via
/// `DateTime.ToLocalTime()`. Block-start flooring (in SessionBlockCalculator) uses UTC
/// instead — that asymmetry is intentional and preserved from the other platforms.
/// </summary>
public static class SnapshotComputer
{
    public static UsageSnapshot ComputeSnapshot(IReadOnlyList<UsageEvent> events, double manualBlockTokenTarget, GaugeAppearance appearance)
    {
        var now = DateTime.UtcNow;
        var nowLocal = now.ToLocalTime();
        var startOfToday = nowLocal.Date;
        var sevenDaysAgo = now.AddDays(-7);

        var todayEvents = events.Where(e => e.Timestamp.ToLocalTime() >= startOfToday).ToList();
        var last7DaysEvents = events.Where(e => e.Timestamp >= sevenDaysAgo).ToList();
        var last7DaysTotalTokens = last7DaysEvents.Sum(e => e.TotalTokens);
        var todayTotalTokens = todayEvents.Sum(e => e.TotalTokens);

        var breakdownByModel = new Dictionary<string, ModelBreakdown>();
        foreach (var evt in todayEvents)
        {
            if (!breakdownByModel.TryGetValue(evt.Model, out var entry))
            {
                entry = new ModelBreakdown { Model = evt.Model };
                breakdownByModel[evt.Model] = entry;
            }
            entry.InputTokens += evt.InputTokens;
            entry.OutputTokens += evt.OutputTokens;
            entry.CacheCreationTokens += evt.CacheCreationTokens;
            entry.CacheReadTokens += evt.CacheReadTokens;
        }

        double todayCost = 0;
        var breakdown = breakdownByModel.Values
            .Where(e => e.InputTokens + e.OutputTokens + e.CacheCreationTokens + e.CacheReadTokens > 0)
            .Select(e =>
            {
                var cost = PricingTable.EstimatedCostUSD(e.Model, e.InputTokens, e.OutputTokens, e.CacheCreationTokens, e.CacheReadTokens);
                e.EstimatedCostUSD = cost;
                todayCost += cost ?? 0;
                return e;
            })
            .OrderBy(e => e.Model, StringComparer.Ordinal)
            .ToList();

        var hourlyTokensToday = HourlyBuckets(todayEvents.Select(e => (e.Timestamp, e.TotalTokens)));
        var dailyTokensLast7Days = DailyBuckets(last7DaysEvents.Select(e => (e.Timestamp, e.TotalTokens)));
        var todayByProject = ProjectBuckets(todayEvents.Select(e => (e.ProjectPath, e.TotalTokens)));

        var blocks = SessionBlockCalculator.ComputeBlocks(events);
        var activeBlock = blocks.Count > 0 && now < blocks[^1].End ? blocks[^1] : null;
        var pastBlocks = activeBlock is null ? blocks : blocks.Take(blocks.Count - 1).ToList();
        long? historicalMaxBlockTokens = pastBlocks.Count > 0 ? pastBlocks.Max(b => b.TotalTokens) : null;

        var blockSummary = activeBlock is null
            ? null
            : new SessionBlockSummary(activeBlock.Start, activeBlock.End, activeBlock.TotalTokens);

        return new UsageSnapshot
        {
            GeneratedAt = now,
            TodayTotalTokens = todayTotalTokens,
            TodayEstimatedCostUSD = todayCost,
            TodayModelBreakdown = breakdown,
            TodayByProject = todayByProject,
            CurrentBlock = blockSummary,
            Last7DaysTotalTokens = last7DaysTotalTokens,
            HourlyTokensToday = hourlyTokensToday,
            DailyTokensLast7Days = dailyTokensLast7Days,
            HistoricalMaxBlockTokens = historicalMaxBlockTokens,
            HistoricalBlockCount = pastBlocks.Count,
            ManualBlockTokenTarget = manualBlockTokenTarget,
            Appearance = appearance
        };
    }

    public static CodexSnapshot ComputeCodexSnapshot(
        IReadOnlyList<CodexUsageEvent> events,
        CodexRateLimitWindow? primaryWindow,
        CodexRateLimitWindow? secondaryWindow,
        string colorHex)
    {
        var now = DateTime.UtcNow;
        var nowLocal = now.ToLocalTime();
        var startOfToday = nowLocal.Date;
        var sevenDaysAgo = now.AddDays(-7);

        var todayEvents = events.Where(e => e.Timestamp.ToLocalTime() >= startOfToday).ToList();
        var last7DaysEvents = events.Where(e => e.Timestamp >= sevenDaysAgo).ToList();
        var last7DaysTotalTokens = last7DaysEvents.Sum(e => e.TotalTokens);
        var todayTotalTokens = todayEvents.Sum(e => e.TotalTokens);

        var breakdownByModel = new Dictionary<string, ModelBreakdown>();
        foreach (var evt in todayEvents)
        {
            if (!breakdownByModel.TryGetValue(evt.Model, out var entry))
            {
                entry = new ModelBreakdown { Model = evt.Model };
                breakdownByModel[evt.Model] = entry;
            }
            entry.InputTokens += evt.InputTokens;
            entry.OutputTokens += evt.OutputTokens;
            entry.CacheCreationTokens += evt.CachedInputTokens;
        }

        var breakdown = breakdownByModel.Values
            .Where(e => e.InputTokens + e.OutputTokens > 0)
            .OrderBy(e => e.Model, StringComparer.Ordinal)
            .ToList();

        var hourlyTokensToday = HourlyBuckets(todayEvents.Select(e => (e.Timestamp, e.TotalTokens)));
        var dailyTokensLast7Days = DailyBuckets(last7DaysEvents.Select(e => (e.Timestamp, e.TotalTokens)));
        var todayByProject = ProjectBuckets(todayEvents.Select(e => (e.ProjectPath, e.TotalTokens)));

        return new CodexSnapshot
        {
            GeneratedAt = now,
            TodayTotalTokens = todayTotalTokens,
            TodayModelBreakdown = breakdown,
            TodayByProject = todayByProject,
            HourlyTokensToday = hourlyTokensToday,
            DailyTokensLast7Days = dailyTokensLast7Days,
            Last7DaysTotalTokens = last7DaysTotalTokens,
            PrimaryWindow = primaryWindow,
            SecondaryWindow = secondaryWindow,
            ColorHex = colorHex
        };
    }

    private static List<HourlyUsagePoint> HourlyBuckets(IEnumerable<(DateTime Timestamp, long Tokens)> events)
    {
        var totalsByHour = new Dictionary<DateTime, long>();
        foreach (var (timestamp, tokens) in events)
        {
            var local = timestamp.ToLocalTime();
            var hourStart = new DateTime(local.Year, local.Month, local.Day, local.Hour, 0, 0, DateTimeKind.Local);
            totalsByHour[hourStart] = totalsByHour.GetValueOrDefault(hourStart) + tokens;
        }
        return totalsByHour
            .Select(kv => new HourlyUsagePoint(kv.Key, kv.Value))
            .OrderBy(p => p.HourStart)
            .ToList();
    }

    private static List<DailyUsagePoint> DailyBuckets(IEnumerable<(DateTime Timestamp, long Tokens)> events)
    {
        var totalsByDay = new Dictionary<DateTime, long>();
        foreach (var (timestamp, tokens) in events)
        {
            var day = timestamp.ToLocalTime().Date;
            totalsByDay[day] = totalsByDay.GetValueOrDefault(day) + tokens;
        }
        return totalsByDay
            .Select(kv => new DailyUsagePoint(kv.Key, kv.Value))
            .OrderBy(p => p.Day)
            .ToList();
    }

    private static List<ProjectUsageEntry> ProjectBuckets(IEnumerable<(string ProjectPath, long Tokens)> events)
    {
        var totalsByProject = new Dictionary<string, long>();
        foreach (var (projectPath, tokens) in events)
        {
            totalsByProject[projectPath] = totalsByProject.GetValueOrDefault(projectPath) + tokens;
        }
        return totalsByProject
            .Select(kv => new ProjectUsageEntry(kv.Key, kv.Value))
            .OrderByDescending(p => p.Tokens)
            .ToList();
    }
}
